### ISSUE TITLE
Missing data from CAAQM

### ISSUE THREAD

**Issue by aatishb** [openaq/openaq-fetch#585]
I’m trying to understand some gaps in the openaq data from CAAQM. For example, if you look at [Anand Vihar](https://openaq.org/#/location/Anand%20Vihar%2C%20Delhi%20-%20DPCC?_k=umo8gq) in Delhi, there’s no data for Wed 19. 

![screen shot 2018-12-23 at 10 17 31 am](https://user-images.githubusercontent.com/1878638/50380940-768c8c80-069f-11e9-8c5d-b3879d6236be.png)

However, if you query [CAAQM](https://app.cpcbccr.com/ccr/#/caaqm-dashboard-all/caaqm-view-data-report/%2522%257B%255C%2522parameter_list%255C%2522%253A%255B%257B%255C%2522id%255C%2522%253A3%252C%255C%2522itemName%255C%2522%253A%255C%2522PM2.5%255C%2522%252C%255C%2522itemValue%255C%2522%253A%255C%2522parameter_193%255C%2522%257D%255D%252C%255C%2522criteria%255C%2522%253A%255C%252215%2520Minute%255C%2522%252C%255C%2522reportFormat%255C%2522%253A%255C%2522Tabular%255C%2522%252C%255C%2522fromDate%255C%2522%253A%255C%252219-12-2018%2520T00%253A00%253A00Z%255C%2522%252C%255C%2522toDate%255C%2522%253A%255C%252219-12-2018%2520T10%253A17%253A59Z%255C%2522%252C%255C%2522state%255C%2522%253A%255C%2522Delhi%255C%2522%252C%255C%2522city%255C%2522%253A%255C%2522Delhi%255C%2522%252C%255C%2522station%255C%2522%253A%255C%2522site_301%255C%2522%252C%255C%2522parameter%255C%2522%253A%255B%255C%2522parameter_193%255C%2522%255D%252C%255C%2522parameterNames%255C%2522%253A%255B%255C%2522PM2.5%255C%2522%255D%257D%2522) directly, you can find data for that day.

I think what might be happening is this: the CAAQM website seems to frequently provide missing entries for current data when queried at 15 minute resolution, so this is probably scraped as missing data. However, they seem to go back and fill in these missing values once they have data on it. So my guess is that openaq might be scraping the initial missing data but not the later update.

**Comment by nishadhka** [comment]
Yes, the URL queried by (openaq)[https://github.com/openaq/openaq-fetch/blob/5f81726ca7646a1824a3a762a2db691387ba1148/adapters/caaqm.js#L63] https://app.cpcbccr.com/caaqms/caaqms_viewdata_v2 gives only current data, for back dated data as in CAAQM website for each station, the url is https://app.cpcbccr.com/caaqms/caaqms_load_delta_v2, may be another adaptor/function with this delta link to run for once in a day to fill missing data.

**Comment by jflasher** [comment]
Hi @aatishb, there are a number of factors which can affect the presence of data in the system. As you  and @nishadhka point out, we're only fetching the current data and if it isn't there when we check back again, we will never get it (the way the system works currently). For this source in particular, we have often seen connectivity issues cause the URL to be inaccessible from our fetching instance which means we'll be missing data then. As @nishadhka points out, we could potentially look to implement a system that does some rectification on a regular basis and tries to fill in gaps. However, this is problematic for at least one big reason. We currently have no concept of QA/QC data points. We store the first value we see for a given location/city/county/parameter. If we see a new data point for that same combo, we do not save it. So if we're going back and getting data a day later, how do we handle this in the system? Do we overwrite the data with new data? Are we storing both and flagging it in the system somehow? And if we're just filling in the gaps, are we potentially presenting QA/QC'd data right alongside real-time data? Figuring out how to better handle this within the system will open us up to many new data sources (historical, QA/QC'd data) which is great! But we currently don't have a good way to handle it.

**Comment by MichalCz** [comment]
@jflasher I think that missing data will be filled in though, won't it? The only thing limiting the insert in the PKEY on `measurements` that would not be there if the old data was not inserted in the first place. In this case perhaps it may be just a matter of extending the lookback window that is used to filter out "old" results?

On the other hand it would create a lot of unnecessary inserts, which wouldn't be good at all. But if we fetched only the index for last 7 days and filter out duplicates without inserting those to DB I don't think we'd have a major increase in execution time (the data is downloaded regardless) and maybe even a small drop?

Now, this filtering could also be extracted to a separate step in fetch so it could be easily extended with some logic later on?

I would suggest writing something like this:

- while sources are selected, start fetching the index for `n` hours back.
- when data starts flowing from adapter, `await` the index for the source
- when index is received filter out the already existent items

**Comment by jflasher** [comment]
@MichalCz How do you think we would best get this index? I'm a little scared of having fetch start hitting the API. But if we could figure out a good way to do this, it'd be interesting to put less work on the database to deny duplicate inserts.

But aside from the technical piece, there is a data similarity question. If the data is QA/QCd and then put up, it's a fundamentally different data quality type than the data we get in real time. I'm not sure a) we know how to tell what's going on during that intervening time when it's non-existent and b) even if we do know it's QA/QC'd, we do not currently have a way to indicate this in our system. For the data to be used in research, that distinction needs to be clear.

**Comment by MichalCz** [comment]
@jflasher I'm thinking of fetching the index beforehand - so when the fetch process starts there could be a query of some sort to fetch the current `lookbackPeriod` set of measurements. Remember this is read and it's a single query and it's fetching an index, so it's cheaper to do.

The query could go in [db.js:222](https://github.com/openaq/openaq-fetch/blob/develop/lib/db.js#L222), and then the check in [db.js:234](https://github.com/openaq/openaq-fetch/blob/develop/lib/db.js#L234), before the `.use(` as something like `.filter(measurement => !dbIndex.includes(mkIndex(measurement)))` 

I'd also suggest the `lookbackPeriod` to go as one of the arguments for the adapters.

**Comment by jflasher** [comment]
This sounds promising, I think you're trying to do something I didn't even know was possible. :) I didn't know it was possible to fetch the index in some sort of useable format? 

If that's possible, this sounds promising.

**Comment by MichalCz** [comment]
I don't think it's possible to get the index in some usable format... not without some trickery...

I'll try some of that trickery...

**Comment by MichalCz** [comment]
Definitely possible. ;) Please check out the PR.

**Comment by aatishb** [comment]
Looking closer at the data, the only current PM 2.5 data in Delhi seems to be the US Embassy. In fact, starting around Nov 20, there have been long stretches ranging over a month altogether, where the only Delhi data source is the Embassy. I'm guessing this is related to the same issue with the CAAQM adapter.
![screen shot 2019-01-11 at 7 58 10 pm](https://user-images.githubusercontent.com/1878638/51039483-694a2a00-15db-11e9-9e31-fdee0ec6c4be.png)

**Comment by RocketD0g** [comment]
Hi @aatishb - I'll just flag another issue is that our IP address appears to get repeatedly blocked by the source and it stops our access until we change the IP address. There are technical ways around this, but our goal ultimately is to have more data being shared as openly as possible at the source, not necessarily to get the data from the source in any way we can or in any way that is 'sneaky.' We welcome anyone who might connect us with the CAAQM to discuss this issue with them. (Also @jflasher please chime in if you have anything to add.)

**Comment by aatishb** [comment]
@RocketD0g thanks for the update. Let me see if I can reach out to someone there.

**Comment by aatishb** [comment]
Hi @RocketD0g, I just heard back a day ago from someone at CPCB. They are asking what ip is being blocked. Is there anyone I should loop in to the email thread? Is there a list of ip addresses I can provide? Thanks.

**Comment by jflasher** [comment]
Hi @aatishb, thanks for following up on this. The IP address that is used for our fetching instance currently is `18.204.140.110`, though that can change. If I use the command below on my personal computer or a similar instance to the fetching instance (but different IP) I receive an HTTP `200` response. 

`curl -kv "https://app.cpcbccr.com/ccr/#/caaqm-dashboard-all/caaqm-landing"`

If I do this on the fetching instance, I receive an HTTP `404` response. And if the IP address of the fetching instance changes, it will work for a day or so before failing again.

This behavior, along with the fact that all access outside of India was blocked previously, leads me to believe that our IP address is being blocked. If there is anyone to talk to about how to best access this data, I'm happy to do that. Please feel free to add jflasher@openaq.org to any email threads you think is appropriate. Thanks!

**Comment by majesticio** [comment]
Closing issue as the CPCB adapter is no longer in use and their new site uses CAPTCHA. New approach is being discussed in #283
